feat(useAuth): add useAuth to @vueuse/firebase

jacobclevenger opened the pull request:
Here is the initial implementation of the `useAuth` function mentioned in issue #309. 

This includes a basic demo with a demo firebase account that users can login to using the demo along with some basic documentation. I tried to follow the way the rest of the docs are written the best I could. Let me know if you missed anything or notice any obvious faults with it.

If you have any questions on the implementation just let me know and I'll do my best to answer!

antfu commented on packages/firebase/useAuth/index.ts:
```diff
@@ -0,0 +1,70 @@
+  createUserWithEmailAndPassword(email: string, password: string): void
+}
+
+export function useAuth(instance: typeof firebase): UseAuth {
```
Since we are only using the `auth`, maybe we can have them pass the auth instance instead?

```ts
const auth = firebase.auth()
const { user, isAuthenticated } = useAuth(auth)
const { signInWithPopup, signOut } = auth

const signin = () => {
  signInWithPopup(new firebase.auth.GoogleAuthProvider())
}
```

antfu replied:
The wrapping for signIn/signOut functions feels a bit heavy to me. Maybe we can leave them as my snippet shows? WDYT?

jacobclevenger replied:
Yeah I think that is a good idea. I ran into an issue with the the `useAuth()` calling `firebaes.auth()` before it was initialized causing an error which took a bit for more to figure out what was going on which  gave me a bit of tunnel vision haha. I think just providing the `user`, `isAuthenticated`, and `authenticationStatus` would be a good idea

jacobclevenger replied:
At this point should I just delete the demo.vue file and just have everything in index.md?

jacobclevenger commented on packages/firebase/useAuth/demo.vue:
```diff
@@ -0,0 +1,63 @@
+if (!firebase.apps.length) {
+  firebase.initializeApp({
+    apiKey: 'AIzaSyBcgy5ezJoYC1Ed7uwstgy9x5pR5nrv6Q8',
+    authDomain: 'vueuse-auth.firebaseapp.com',
```
Yeah it was just one I created for the demo, can easily be changed if you'd like

antfu replied:
This could potentially bring some privacy issues, are you good with moving the code to the `readme.md` to have a "Full demo" code snippet instead of runnable demo? Or do you have any better solution for it?

jacobclevenger replied:
Agreed, I'll delete the firebase app and we can move it to just the demo. Although I don't know if it is that much of a privacy concern as these are things that would normally be shipped in a client side app I believe and they are restricted by domains. Either way just for simplicity we can move the demo to the docs

jacobclevenger commented:
Hey @antfu I have simplified the API and removed the demo.vue and moved the example it into the readme. Let me know what you think of these changes.

Sorry for the messy process, this is my first time contributing a feature to an open-source project 😅

antfu commented on packages/firebase/useAuth/index.ts:
```diff
@@ -0,0 +1,22 @@
+}
+
+export function useAuth() {
+  const { auth } = firebase
```
Is that safe to use global instance? Or should we at least provide an option to override it?

jacobclevenger replied:
What about something like this snippet below? I ran into an issue earlier where I was calling `firebase.init` in the same component that I was calling useAuth, but as long as they call `firebase.init` in something like `main.js` before `useAuth` is used, it shouldn't be a problem. I guess maybe the only problem might be if they use the `useAuth` in the main.js file then it could become a problem The snippet below would also allow the user to provide a non-global instance of firebase auth, but if they don't it will just fallback to the global instance.

```ts
export function useAuth(authInstance?: typeof firebase.auth) {
  const auth = authInstance ? authInstance() : firebase.auth()

  const user = ref<firebase.User | null>(auth.currentUser)
  const isAuthenticated = computed(() => !!user.value)

  auth.onIdTokenChanged(authUser => user.value = authUser)

  return {
    isAuthenticated,
    user,
  }
}
```

antfu replied:
Yeah, that's much better. But instance of authInstance, it's probably better to be authConstructor? Also think it might be possible to support both constructor and instance?

jacobclevenger replied:
Oh so you mean something like supporting the use of both `useAuth(firebase.auth)` and `useAuth(firebase.auth())`?

If that is what you mean then I'll update make an update tomorrow to support that :)

antfu replied:
Exactly. Thanks.

jacobclevenger replied:
Sorry it took a while, got busy with some stuff. Just implemented it though, let me know what you think 😅

antfu commented on packages/firebase/useAuth/index.ts:
```diff
@@ -0,0 +1,22 @@
+export function useAuth() {
+  const { auth } = firebase
+
+  const user = ref<firebase.User | null>(auth().currentUser)
```
Is calling is sync, it looks like there is not different to call it at the beginning of the setup(). As long as you do the initialization in main.js I thought it should be fine. What's the problem you are running into?

antfu commented:
Looking good, thanks!
